### ISSUE TITLE
Automatic redirection with codes 3xx looks unsafe

### ISSUE THREAD

**Issue by miketsts** [yhirose/cpp-httplib#430]
As I can see, if a client receives a code 3xx and `follow_location_` is true, the request is redirected automatically as it is. I think this treatment should be somewhat related to a specific code.

In my case, the server responds to a POST request with code 303 (See Other) to inform that the resource already exists. In this case, I don't need redirection at all. While GET in such case could be redirected safely.
Of course, it's not a critical problem, as `follow_location_` is `false` by default. So it's just an improvement suggestion.

According to [RFC7231](https://tools.ietf.org/html/rfc7231#section-6.4.4):
> This status code is applicable to any HTTP method.  It is primarily
   used to allow the output of a POST action to redirect the user agent
   to a selected resource, since doing so provides the information
   corresponding to the POST response in a form that can be separately
   identified, bookmarked, and cached, independent of the original
   request

For codes 301, 302 and 307 the RFC contains _"...The user agent MAY use the Location field value for automatic redirection..."_, which well corresponds having the boolean `follow_location_` flag. At the same time, for 302 it has also 
> Note: For historical reasons, a user agent MAY change the request
      method from POST to GET for the subsequent request.  If this
      behavior is undesired, the 307 (Temporary Redirect) status code
      can be used instead.

To sum up, I suggest to keep automatic redirection only for codes 301, 302 and 307 - and don't do it in other cases. I don't know if there is a danger of compatibility break. I hope very much the correct usage of the codes by users keeps the risk very low.

I really enjoy using your library :-). It's pretty simple but powerful at the same time.

**Comment by yhirose** [comment]
@miketsts, thank you for your feedback. I did some research on the RFC7231 and the curl implementation. But I still think there is no problem to redirect for 303 as well.

The first sentence of PRC7231 section 6.4.4 says: 
> The 303 (See Other) status code indicates that the server is **redirecting** the user agent to a different resource, as indicated by a URI in the Location header field, which is intended to provide an  indirect response to the original request.

Also I tried to make `follow_location_` same as [curl_easy_setopt API](https://curl.haxx.se/libcurl/c/CURLOPT_FOLLOWLOCATION.html) in curl. The description of this API says:
> A long parameter set to 1 tells the library to follow any Location: header that the server sends as part of an HTTP header in a **3xx** response. The Location: header can specify a relative or an absolute URL to follow.

So there is no distinction between 303 and other 3xx.

But I found a problem in the current cpp-httplib implementation which is different from RFC7231 and the curl `curl_easy_setopt` function.

> A user agent can perform  a retrieval request targeting that URI (a **GET or HEAD** request if using HTTP), which might also be redirected, and present the eventual result as an answer to the original request

https://github.com/curl/curl/blob/741cb81d09db781c975ca263203befd01ccd4840/lib/transfer.c#L1719-L1728

So the client needs to use `GET` when redirecting. (The curl disable this behavior by applying `CURL_REDIR_POST_ALL` or `CURL_REDIR_POST_303`). I am going to make the same behavior in cpp-httplib when I have time.

Anyway, thank you for your interest in this library and fine report!

**Comment by miketsts** [comment]
@yhirose please note the following in the RFC7231:

> A 303 response to a GET request...  ...the Location field
   value refers to a resource that is descriptive of the target
   resource, such that making a retrieval request on that other resource
   might result in a representation that is useful to recipients without
   implying that it represents the original target resource.

Note "without implying" clause which, in my opinion, also prevents automatic redirection.
As for non-idempotent requests like POST, automatic redirection seems just dangerous as it might result in changes in the location which was not intended by the caller.

There is [Wikipedia page for HTTP 303](https://en.wikipedia.org/wiki/HTTP_303) which, in particular, mentions:
>  This status code should be used with the location header, as described below. If a server responds to a POST or other non-idempotent request with a 303 See Other response and a value for the location header, the client is expected to obtain the resource mentioned in the location header using the GET method; to trigger a request to the target resource using the same method, the server is expected to provide a 307 Temporary Redirect response. 

Still, in my opinion, automatic replacement from POST to (redirected) GET seems inappropriate. In particular, for example, the expected result from POST is 201, while GET usually responds 200, not talking about other response data.

**Comment by yhirose** [comment]
@miketsts, you might be correct, but curl doesn't behave like you explains... Here is my test result with curl:

server.cc
```cpp
#include <httplib.h>
using namespace httplib;

int main(void) {
  Server svr;

  svr.Post("/1", [](const Request & /*req*/, Response &res) {
    std::cout << "POST: /1" << std::endl;
    res.set_redirect("/2");
    res.status = 303;
  });

  svr.Get("/2", [](const Request & /*req*/, Response &res) {
    std::cout << "GET: /2" << std::endl;
    res.set_content("Hello World!", "text/plain");
  });

  svr.listen("localhost", 8080);
}
```

curl output
```bash
$ curl http://localhost:8080/1 --data "body" -v
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> POST /1 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 4
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 4 out of 4 bytes
< HTTP/1.1 303 See Other
< Content-Length: 0
< Location: /2
< 
* Connection #0 to host localhost left intact
* Closing connection 0

$ curl http://localhost:8080/1 --data "body" -v -L
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
> POST /1 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.64.1
> Accept: */*
> Content-Length: 4
> Content-Type: application/x-www-form-urlencoded
> 
* upload completely sent off: 4 out of 4 bytes
< HTTP/1.1 303 See Other
< Content-Length: 0
< Location: /2
< 
* Connection #0 to host localhost left intact
* Issue another request to this URL: 'http://localhost:8080/2'
* Disables POST, goes with GET
* Found bundle for host localhost: 0x7fc9366184c0 [can pipeline]
* Could pipeline, but not asked to!
* Re-using existing connection! (#0) with host localhost
* Connected to localhost (::1) port 8080 (#0)
> GET /2 HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.64.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Length: 12
< Content-Type: text/plain
< 
* Connection #0 to host localhost left intact
Hello World!* Closing connection 0

```

server output
```bash
$ ./server
POST: /1
POST: /1
GET: /2
```

As you see the above, curl behaves exactly same as I explained. It might not be correct according to RFC7231 though, I would like to make cpp-httplib as curl does, because curl behavior is like a de facto standard for HTTP client. (`set_follow_location(true)` is equivalent to the curl `-L` option)

@PixlRainbow, do you have any opinion about this issue?

**Comment by yhirose** [comment]
The cpp-httplib client now behaves like curl for 303 response to a POST request. (#431)

https://github.com/yhirose/cpp-httplib/blob/c2b6e4ac044759db71cbd5407904168e6a1e3d44/test/test.cc#L752-L758

https://github.com/yhirose/cpp-httplib/blob/c2b6e4ac044759db71cbd5407904168e6a1e3d44/test/test.cc#L1333-L1346

**Comment by yhirose** [comment]
@PixlRainbow, do you have any opinion about this issue?
